### ISSUE TITLE
[Proposal] Mutliple For ... Next and Multiple ForEach ... Next

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#48]
## `For ... Next` Loop
Given the following section of code, it can be written as
```vbnet
For x = 0 To 9
  For y = 0 To 9
    Console.WriteLine($"{x} * {y} = {x*y}")
  Next
Next
```
or
```vbnet
For x = 0 To 9
  For y = 0 To 9
    Console.WriteLine($"{x} * {y} = {x*y}")
Next x. y
```
Yet we can not write
```vbbet
For x = 0 To 9, y = 0 To 9
    Console.WriteLine($"{x} * {y} = {x*y}")
Next x. y
```
or
```vbnet
For x = 0 To 9, y = 0 To 9
    Console.WriteLine($"{x} * {y} = {x*y}")
Next '  equivalent to Next y : Next x
```
We should enable the last two scenario to compile.
**Note:** For what if the inner code contains `exit` `continue` `

--------

## `For Each ... Next` Loop
**Premise**
The premise of this feature is to allow the user to specify multiple iterators on a single `For Each`.

**Basic Structrure**
```
For Each x In xs, y In ys
   ' ...
   ' exit       ' exits ys iterator
   ` exit x     ' exits xs iterator
   ' continue   ' goto next iteration of ys
   ' continue x ' goto next iteration of xs
   ' Next       ' an error? / goto next iteration of ys then xs?
Next y , x
```

Example

```vbnet
Public Overrides ReadOnly Property AdditionalLocations As IReadOnlyList(Of Location)
  Get
    Dim builder = ArrayBuilder(Of Location).GetInstance()
    For Each sym In AmbiguousSymbols, l In sym.Locations
      builder.Add(l)
    Next l, sym
    Return builder.ToImmutableAndFree()
  End Get
End Property
```
semantically equivalent to the following
```vbnet
Public Overrides ReadOnly Property AdditionalLocations As IReadOnlyList(Of Location)
  Get
    Dim builder = ArrayBuilder(Of Location).GetInstance()
    For Each sym In AmbiguousSymbols
      For Each l In sym.Locations
        builder.Add(l)
      Next
    Next
    Return builder.ToImmutableAndFree()
  End Get
End Property
```
~**Edit** `For Each` already supports this syntax.~

**Comment by zspitz** [comment]
My own feeling regarding `Exit`, `Continue` and `Next` without a specifier, is that it should apply to the entire loop body. Even though the mental model here is of multiple nested loops, the code as is written contains only a single loop body.

**Comment by Bill-McC** [comment]
Doesn't seem to be adding anything, just skipping a line. As this is to do with iterating, doesn't LINQ address this ?

**Comment by DWAK-ATTK** [comment]
I agree w/ Bill-McC.  This doesn't appear to be adding anything of substantial value to the language.  If you want a language designed for brevity, there's C#.  Personally I like the verbosity of VB and would oppose any feature designed solely for the purpose of removing code clarity.

**Comment by zspitz** [comment]
@DWAK-ATTK I would suggest that a better reason to prefer VB.NET over C# is not because C# was designed for brevity, but because C#'s brevity is in it's usage of symbols. From that perspective, there is no reason for VB.NET to discourage less/smaller syntax, as long as that doesn't come at the cost of a hodgepodge of curly braces and angle brackets.. (NB. are there any official documents / links describing the design goals of VB.NET?)

As an example of smaller syntax: VB.NET has a `For ... Each` syntax even though you could express this using calls to the `GetEnumerator` interface methods.

Also, there is a limit to how far we can go to avoid C#-like brevity with symbols. Mathematical operations, assignment, value equality testing and string formatting are all expressed using symbols in VB.NET. If the goal is to avoid symbols entirely, why don't we have `Plus`, `Minus`, `Times` and `DividedBy` operators?

---

> would oppose any feature designed solely for the purpose of removing code clarity.

Are you saying that this (currently supported syntax):
```
For Each asm In AppDomain.CurrentDomain.GetAssemblies
    For Each t In asm.GetTypes
        For Each m In t.GetMethods
            Console.WriteLine(m.Name)
Next asm, t, m
```
is clearer / more useful than this?
```
For Each asm In AppDomain.CurrentDomain.GetAssemblies, t In asm.GetTypes, m in t.GetMethods
    Console.WriteLine(m.Name)
Next
```

**Comment by DWAK-ATTK** [comment]
@zspitz 
You put words in my mouth and then ask me to defend them.  I will not.  If you wish to address the words that I did say then I am willing to discuss that.

As for the question you pose regarding the example of currently supported syntax vs proposed syntax: In my opinion the currently supported syntax is far superior, though I prefer matching Next statements for each For statement.

Having separate For statements makes it easy to set up processing at each loop "level".  Ditto for matching Next statements.

Allowing the proposed syntax means that a single For statement could kick off multiple nested loops.  For a single statement to do that doesn't sit well with me.  As I visually scan through the code base I expect, to the degree possible, each statement to do a single task.

**Comment by Bill-McC** [comment]
@zspitz 

> Are you saying that this (currently supported syntax):
```
For Each asm In AppDomain.CurrentDomain.GetAssemblies
    For Each t In asm.GetTypes
        For Each m In t.GetMethods
            Console.WriteLine(m.Name)
Next asm, t, m

```
Won't compile. You'll be forced to put the m, t, asm in the correct order, or omit them.

**Comment by AdamSpeight2008** [comment]
@DWAK-ATTK 
LINQ query statements permits multiple "loops" in a single "statement" 
```vbnet
Module Module1
  Public Sub Main()
    Dim xs = From x In (0).To(9), y In (0).To(9) Select $"{x} * {y} = {x * y}"
    For Each res In xs
      Console.WriteLine(res)
    Next

  End Sub

  <Runtime.CompilerServices.Extension>
  Iterator Function [To]([from] As Integer, upto As Integer) As IEnumerable(Of Integer)
    For value = [from] To upto
      Yield value
    Next
  End Function
End Module
```
If we are allow to write combine multiple `next`'s into a single statement, we should be also be allow to combine multiple `For`'s. We shouldn't dictating what style one "should be" using, that "must be" left for the user to decide.

**Comment by AdamSpeight2008** [comment]
@DWAK-ATTK  The query syntax does also permit 
```vb.net
For Each asm In AppDomain.CurrentDomain.GetAssemblies,
	   t In asm.GetTypes,
	   m In t.GetMethods
    Console.WriteLine(m.Name)
Next
```

**Comment by Bill-McC** [comment]
@AdamSpeight2008  The query syntax requires a "From" . Not sure what that sample is, but it isn't LINQ.

And I think that really adds to the reasons not to do this:
(a) we already have alternative LINQ syntax,
(b) doesn't add any new functionality
(c) could be confusing with query syntax as to whether or not it is an expression or multiple statements

**Comment by AdamSpeight2008** [comment]
@Bill-McC Since the introduction of implicit line continuations and multi-line lambdas, the line between expression and multiple statements is a little blurry.

**Comment by AnthonyDGreen** [comment]
@AdamSpeight2008

I strongly agree with this. There's a strong precedent in the language that multiple statements (and query clauses) can be combined into one comma separated list. We see this with Imports, field and variable declarations, the From and Let query clauses, the Next statement as you point out, the Using statement, the Case statement, the Inherits and Implements statement, and probably a few more I'm forgetting. I think this idea is completely consistent with the existing syntactic conventions of the language.

**Comment by Bill-McC** [comment]
@AdamSpeight2008  Still can't get that code to compile: shows as syntax error. Are you sure it works ? What version of VB ?

```
For Each asm In AppDomain.CurrentDomain.GetAssemblies,
	   t In asm.GetTypes,
	   m In t.GetMethods
    Console.WriteLine(m.Name)
Next
```

**Comment by Bill-McC** [comment]
@AnthonyDGreen  Strongly disagree. This comma list proposed dictates a specific order; and outer and inner. The only place that exists that I can think of in VB is within LINQ and I suggest it stays confined to LINQ.

**Comment by AdamSpeight2008** [comment]
@Bill-McC  `Next` dictates a specific order.

**Comment by AdamSpeight2008** [comment]
@Bill-McC I don't know, but it compiled and ran.

**Comment by Bill-McC** [comment]
@AdamSpeight2008   I still can't get it to compile

**Comment by Bill-McC** [comment]
@AdamSpeight2008  Next ... does not*dictate* a specific order, it has to comply with the specific order dictated by the For loops.

**Comment by AdamSpeight2008** [comment]
@Bill-McC You can specific a specific order, whether it'll compile is a different matter., 

Also you say it `does not *dictate* a specific order`, than later it does. Which is it? Or are just being a troll?

**Comment by Bill-McC** [comment]
@AdamSpeight2008  Suggest you re-read what I wrote: it was specific. Also suggest you keep the conversation to either "proposed" syntax or syntax that actually compiles, not base arguments on syntax that doesn't compile today as proof of some existence. 
And please cut the crap -> as in don't call people trolls because you can't answer simple questions about code you says compiles which doesn't.
And can you please stop with the regular opening/closing of this suggestion.  Thanks.

**Comment by AdamSpeight2008** [comment]
@Bill-McC I think you need to learn the difference of I allowed by grammar / syntax.  And what it a valid compilation.

```vb.net
Imports System
Public Class C
    Public Sub M()
        For x = 0 To 10
        For y = 0 to 10
        Console.WriteLine($"x: {x} y: {y} = {x*y}")
    Next x,y
    End Sub
End Class
``` 

[View on SharpLab](https://sharplab.io/#b:master/f:vb/JIWwDg9gTgLgzgAgMoE84wKYgLACgAKArgEYA2AlgMYIDCpAhnIjXgmwkWVciQgLIAKAJSt2YgGLQEADwQBeBAAYEAFQgIAjItFi2kqAhTylCGOq07dNCADs4EUhgB0AdSjlMAGXI2MAgCQARNIAXAgA3tIAvoZh4SgxCpEAVAmBIrhiAHIY0jAyADQolgCiNgAmPMR4ZZV0jHBAA===) allowed by the syntax. but produces the error (from the binding stage *I believe*

**Comment by Bill-McC** [comment]
@AdamSpeight2008  If the language specification says it causes a compilation error, it is not valid code.
As I said earlier, I don't see the point in trying to base arguments for a syntax based on a syntax that doesn't compile.  The language spec is clear on this:

>  A Next statement without a variable matches the innermost open For statement, while a Next statement with one or more loop control variables will, from left to right, match the For loops that match each variable. If a variable matches a For loop that is not the most nested loop at that point, a compile-time error results.

Now you still have not answered the previous questions about the code you said compiled and ran. I see you have edited the original proposal, so I'll take that as you admit it was incorrect. Please be aware other people won't see that clearly, so it would be nice for the sake of clarity if you indicated it.

And finally, again I ask you to stop the crap. Previously you said I was trolling, now you tell me what I need to learn. I could retort, but you aren't worth it. Just stop, okay ?

**Comment by AdamSpeight2008** [comment]
@Bill-McC I already answered that. 

Where in the syntax does it prevent it?
```
ForStatement
    : 'For' LoopControlVariable Equals Expression 'To' Expression
      ( 'Step' Expression )? StatementTerminator
      Block?
      ( 'Next' NextExpressionList? StatementTerminator )?
    ;

LoopControlVariable
    : Identifier ( IdentifierModifiers 'As' TypeName )?
    | Expression
    ;

NextExpressionList
    : Expression ( Comma Expression )*
    ;
```
Which is allows.
```
For x = 0 To 10
For y = 0 To 10
  ...
Next x , y
```
We you dictating `x , y' as the loop evaluation order. It the language semantics that cause it to me invalid, which is that quotation from the specification.

**Comment by DualBrain** [comment]
The Next in all of the above examples are syntactically wrong:

`For x = 1 To 5
  For y = 1 To 5
    ...
Next y, x`

The order of the variables on the Next statement must be in the proper order and is checked for accuracy.

In other words, the ability to have the variable on the Next statement is there for those that are using the variable being there as an additional protection/validation mechanism; not for flow control.

Because of this, I think there is a bit of confusion regarding the suggestion.

I think it's something interesting to think about with regards to enhancing the For portion to allow for multiple loops as a single statement.  With that said, I think it becomes difficult to the user/reader to understand intuitively which loop runs as the inner versus outer.  One could argue that it's just shorthand for the long hand version being that the first loop listed is the outermost loop and each loop on the same line is run within the outermost loop defined just prior.

The necessity to have the variable name on the Next statement should also be optional (as it is today - in other words, absolutely no change).

With all of that said, I'd like to suggest that this only be something further pursued if there is minimal effort involved to implement as this doesn't necessarily gain us any huge increase in productivity (in my mind) over what currently exists.  

I also want to state that I'm not a fan of trying to modify the Next statement to support the ability to "control execution".  That, I believe, would ultimately create nothing but a huge mess.  Leave it as a "validation" / "shorthand" mechanism as it is today.

`For x = 1 To 100, y = 10 To 100
   ...
Next y, x`

versus

`For x = 1 To 100 : For y = 10 To 100
   ...
Next y, x`

This saves 3 keystrokes...

Furthermore, when working with this single type loop, it does create additional problems with regards to how (if the need should arise) where you'd like to control the flow of the code between the different loops.

` For x = 1 To 100 : For y = 10 To 100
  Console.WriteLine($"{x}, {y}")
  Exit For
Next y, x`

The above code will execute the full loop (without any ability to break out of) for x as the Exit For only works for the "inner loop".  One could break out of this using a Goto...

I see that #186 discusses (somewhat) this side of the problem.

**Comment by AnthonyDGreen** [comment]
**Update**: The [12/6/2017](https://github.com/dotnet/vblang/blob/master/meetings/2017/vbldm-notes-2017.12.06.md) LDM approved this idea in principle. The `For Each` bit will fall out of doing #104 but the `For` part still seems like a very reasonable and natural extension to the language.
